narwhals/functions.py

```diff
@@ -430,6 +432,163 @@ def _from_dict_impl(
+            necessary if inputs are not Narwhals Series.
+
+    Returns:
+        A new DataFrame
+
+    Examples:
+        >>> import pandas as pd
+        >>> import polars as pl
+        >>> import pyarrow as pa
+        >>> import narwhals as nw
+        >>> import numpy as np
+        >>> from narwhals.typing import IntoFrameT
+        >>> data = {"a": [1, 2], "b": [3, 4]}
+
+        Let's create a new dataframe of the same class as the dataframe we started with, from a dict of new data:
+
+        >>> def agnostic_from_numpy(df_native: IntoFrameT) -> IntoFrameT:
+        ...     new_data = np.array([[5, 2, 1], [1, 4, 3]])
+        ...     df = nw.from_native(df_native)
+        ...     native_namespace = nw.get_native_namespace(df)
+        ...     return nw.from_numpy(new_data, native_namespace=native_namespace).to_native()
+
+        Let's see what happens when passing pandas, Polars or PyArrow input:
+
+        >>> agnostic_from_numpy(pd.DataFrame(data))
+           column_0  column_1
+        0         5         1
+        1         2         4
+        2         1         3
+        >>> agnostic_from_numpy(pl.DataFrame(data))
+        shape: (3, 2)
+        ┌──────────┬──────────┐
+        │ column_0 ┆ column_1 │
+        │ ---      ┆ ---      │
+        │ i64      ┆ i64      │
+        ╞══════════╪══════════╡
+        │ 5        ┆ 1        │
+        │ 2        ┆ 4        │
+        │ 1        ┆ 3        │
+        └──────────┴──────────┘
+        >>> agnostic_from_numpy(pa.table(data))
+        pyarrow.Table
+        column_0: int64
+        column_1: int64
+        ----
+        column_0: [[5,2,1]]
+        column_1: [[1,4,3]]
+    """
+    from narwhals import dtypes
+
+    return _from_numpy_impl(
+        data,
+        schema,
+        native_namespace=native_namespace,
+        dtypes=dtypes,  # type: ignore[arg-type]
+    )
+
+
+def _from_numpy_impl(
+    data: np.ndarray,
+    schema: dict[str, DType] | Schema | None = None,
+    *,
+    native_namespace: ModuleType | None = None,
+    dtypes: DTypes,
+) -> DataFrame[Any]:
+    if data.ndim != 2:
+        msg = "`from_numpy` only accepts 2D numpy arrays"
+        raise ValueError(msg)
+    if native_namespace is None:
+        msg = "Calling `from_numpy` without `native_namespace` is not supported"
+        raise TypeError(msg)
+    implementation = Implementation.from_native_namespace(native_namespace)
+
+    if implementation is Implementation.POLARS:
+        if schema:
+            from narwhals._polars.utils import (
+                narwhals_to_native_dtype as polars_narwhals_to_native_dtype,
+            )
+
+            schema_pl = {
+                name: polars_narwhals_to_native_dtype(dtype, dtypes=dtypes)
+                for name, dtype in schema.items()
+            }
+            native_frame = native_namespace.from_numpy(data, schema=schema_pl)
+        else:
+            native_frame = native_namespace.from_numpy(data, orient="col")
```

Comment:
Why are we defaulting to `orient="col"`? To me this seems the opposite of what polars does if nothing else is provided:

```py
import numpy as np
import polars as pl

data = np.array([[5, 2, 1], [1, 4, 3]])
data
array([[5, 2, 1],
       [1, 4, 3]])

pl.from_numpy(data)
shape: (2, 3)
┌──────────┬──────────┬──────────┐
│ column_0 ┆ column_1 ┆ column_2 │
│ ---      ┆ ---      ┆ ---      │
│ i64      ┆ i64      ┆ i64      │
╞══════════╪══════════╪══════════╡
│ 5        ┆ 2        ┆ 1        │
│ 1        ┆ 4        ┆ 3        │
└──────────┴──────────┴──────────┘
```

while:

```py
pl.from_numpy(data, orient="col")
shape: (3, 2)
┌──────────┬──────────┐
│ column_0 ┆ column_1 │
│ ---      ┆ ---      │
│ i64      ┆ i64      │
╞══════════╪══════════╡
│ 5        ┆ 1        │
│ 2        ┆ 4        │
│ 1        ┆ 3        │
└──────────┴──────────┘
```

This translates to the array transpose in pandas

Reply:
Hey @FBruzzesi, thanks for your review! I fixed the orientation from columns to rows